crypto/fipsmodule/modes/asm/aesv8-gcm-armv8-unroll8.pl

```diff
@@ -0,0 +1,7380 @@
+	ldr	$rk10q, [$cc, #160]					@ load rk10
+
+	aese	$ctr3b, $rk9						@ AES block 8k+11 - round 9
+	aese	$ctr4b, $rk8  \n  aesmc	$ctr4b, $ctr4b			@ AES block 8k+12 - round 8
+	aese	$ctr2b, $rk9						@ AES block 8k+10 - round 9
+
+	aese	$ctr5b, $rk8  \n  aesmc	$ctr5b, $ctr5b			@ AES block 8k+13 - round 8
+	aese	$ctr1b, $rk8  \n  aesmc	$ctr1b, $ctr1b			@ AES block 8k+9 - round 8
+	aese	$ctr6b, $rk9						@ AES block 8k+14 - round 9
+
+	aese	$ctr4b, $rk9						@ AES block 8k+12 - round 9
+	add	$main_end_input_ptr, $main_end_input_ptr, $input_ptr
+	aese	$ctr0b, $rk9						@ AES block 8k+8 - round 9
+
+	aese	$ctr7b, $rk9						@ AES block 8k+15 - round 9
+	aese	$ctr5b, $rk9						@ AES block 8k+13 - round 9
+	aese	$ctr1b, $rk9						@ AES block 8k+9 - round 9
+
+	add	$end_input_ptr, $input_ptr, $bit_length, lsr #3		@ end_input_ptr
+	cmp	$input_ptr, $main_end_input_ptr				@ check if we have <= 8 blocks
+	b.ge	.L128_enc_tail						@ handle tail
+
+	ldp	$ctr_t0q, $ctr_t1q, [$input_ptr], #32			@ AES block 0, 1 - load plaintext
+
+	ldp	$ctr_t2q, $ctr_t3q, [$input_ptr], #32			@ AES block 2, 3 - load plaintext
+
+	ldp	$ctr_t4q, $ctr_t5q, [$input_ptr], #32			@ AES block 4, 5 - load plaintext
+
+	ldp	$ctr_t6q, $ctr_t7q, [$input_ptr], #32			@ AES block 6, 7 - load plaintext
+	cmp	$input_ptr, $main_end_input_ptr				@ check if we have <= 8 blocks
+
+	eor3	$res0b, $ctr_t0b, $ctr0b, $rk10				@ AES block 0 - result
+	rev32	$ctr0.16b, $rtmp_ctr.16b				@ CTR block 8
+	add	$rtmp_ctr.4s, $rtmp_ctr.4s, $rctr_inc.4s		@ CTR block 8
+
+	eor3	$res1b, $ctr_t1b, $ctr1b, $rk10				@ AES block 1 - result
+	stp	$res0q, $res1q, [$output_ptr], #32			@ AES block 0, 1 - store result
+
+	rev32	$ctr1.16b, $rtmp_ctr.16b				@ CTR block 9
+	eor3	$res5b, $ctr_t5b, $ctr5b, $rk10				@ AES block 5 - result
+	add	$rtmp_ctr.4s, $rtmp_ctr.4s, $rctr_inc.4s		@ CTR block 9
+
+	eor3	$res2b, $ctr_t2b, $ctr2b, $rk10				@ AES block 2 - result
+	eor3	$res6b, $ctr_t6b, $ctr6b, $rk10				@ AES block 6 - result
+	eor3	$res4b, $ctr_t4b, $ctr4b, $rk10				@ AES block 4 - result
+
+	rev32	$ctr2.16b, $rtmp_ctr.16b				@ CTR block 10
+	add	$rtmp_ctr.4s, $rtmp_ctr.4s, $rctr_inc.4s		@ CTR block 10
+
+	eor3	$res3b, $ctr_t3b, $ctr3b, $rk10				@ AES block 3 - result
+	eor3	$res7b, $ctr_t7b, $ctr7b,$rk10				@ AES block 7 - result
+	stp	$res2q, $res3q, [$output_ptr], #32			@ AES block 2, 3 - store result
+
+	rev32	$ctr3.16b, $rtmp_ctr.16b				@ CTR block 11
+	add	$rtmp_ctr.4s, $rtmp_ctr.4s, $rctr_inc.4s		@ CTR block 11
+	stp	$res4q, $res5q, [$output_ptr], #32			@ AES block 4, 5 - store result
+
+	stp	$res6q, $res7q, [$output_ptr], #32			@ AES block 6, 7 - store result
+
+	rev32	$ctr4.16b, $rtmp_ctr.16b				@ CTR block 12
+	add	$rtmp_ctr.4s, $rtmp_ctr.4s, $rctr_inc.4s		@ CTR block 12
+	b.ge	.L128_enc_prepretail					@ do prepretail
+
+.L128_enc_main_loop:							@ main loop start
+	rev32	$ctr5.16b, $rtmp_ctr.16b				@ CTR block 8k+13
+	ldr	$h5q, [$current_tag, #128]				@ load h5l | h5h
```

Comment:
Can we pair all the `ldr` instructions?  For instance:
```
ldp $h5q, $h56kq, [$current_tag, #128]
```
Also, we should issue all the loads first and then do the `ext`.

Reply:
sebpop offline:
> I can see that there's no register pressure, so having the loads as early as possible is a good thing.
Could you please try to move all the loads up at the beginning of the loop, the loads can be in ascending order: (offsets +0, +16, +32, ...), followed by the ext instructions, and then the computation.

I tried that in Decrypt (not in Encrypt yet). It did remove the regression when it was applied to the higher hash registers: h5, h56, h6, h7, h78, h8.
Strangely, when I applied it to the lower hash registers, h1, h12, h2, h3, h34, h4, the regression came back.
So if I don't apply the last 3 hunks in the attached file, there is no regression in Decrypt (but no improvement, either).
[ldp-current_tag-dec-regress.patch](https://github.com/aws/aws-lc/files/11113011/ldp-current_tag-dec-regress.patch)

Reply:
Thanks Nevine.
It does not seem to bring much difference by pairing the loads and hoisting them to the beginning of the loop. 
Let's keep the current patch without the change.